peer.go

```diff
@@ -454,6 +424,130 @@ func (p *peer) loadActiveChannels(chans []*channeldb.OpenChannel) error {
 	return nil
 }
 
+// addLink creates and adds a new link from the specified channel.
+func (p *peer) addLink(chanPoint *wire.OutPoint,
+	lnChan *lnwallet.LightningChannel,
+	forwardingPolicy *htlcswitch.ForwardingPolicy,
+	blockEpoch *chainntnfs.BlockEpochEvent,
+	chainEvents *contractcourt.ChainEventSubscription,
+	currentHeight int32, syncStates bool) error {
+
+	// onChannelFailure will be called by the link in case the channel
+	// fails for some reason.
+	onChannelFailure := func(chanID lnwire.ChannelID,
+		shortChanID lnwire.ShortChannelID,
+		linkErr htlcswitch.LinkFailureError) {
+
+		// The link has notified us about a failure. We launch a go
+		// routine to stop the link, disconnect the peer and optionally
+		// force close the channel. We must launch a goroutine since we
+		// must let OnChannelFailure return in order for the link to
+		// completely stop in the call to RemoveLink.
+		p.wg.Add(1)
+		go func() {
+			defer p.wg.Done()
+
+			// We begin by removing the link from the switch, such
+			// that it won't be attempted used for any more
+			// updates.
+			err := p.server.htlcSwitch.RemoveLink(chanID)
+			if err != nil {
+				peerLog.Errorf("unable to stop link(%v): %v",
+					shortChanID, err)
+			}
+
+			// If the error encountered was severe enough, we'll
+			// now force close the channel.
+			if linkErr.ForceClose {
+				peerLog.Warnf("Force closing link(%v)",
+					shortChanID)
+
+				closeTx, err := p.server.chainArb.ForceCloseContract(*chanPoint)
```

Comment:
Might we want this to execute before spawning the goroutine, to prevent any possibility of the link being readded to the switch before this executes? I think there could be interleaving that could permit this if the remote party disconnects and reconnects super quick, and the db still shows the channel as open.

Granted, it would likely be difficult in practice :P

Would this cause any races you're aware of with the contract court?

Reply:
This was attempted at first, but it will cause a deadlock as the `ChannelArbitrator` will call `RemoveLink`, which in turns call `link.Stop()`.

Maybe this won't be a problem since we force close before we disconnect the peer?

Reply:
Well it's safe if we assume that we're the only ones who can disconnect, but the issue i'm thinking of happens when the remote peer disconnects, which we can't control

Reply:
Alright, verdict on this was delay the untangling of switch/chainarb/peer life cycle around force close to a later PR. We will circle back and lock everything down in a follow up.